### PR TITLE
Adds option to disable filelock, updates tests (Issue #90)

### LINKED ISSUE #90: Multiprocessing file lock hangs on certain filesystems
I recently noticed that cppimport version 22.7.12+ hangs during the build process (our computing center NERSC recently reconfigured their filesystem). The culprit turned out to be a file lock in  introduced in #67 and merged in #71 , which enables multiple processes to use a cppimport-compiled module without stepping on each other's toes. The lock is on line `https://github.com/tbenthompson/cppimport/blob/main/cppimport/importer.py#L49`. 

Unfortunately, our new filesystem configuration does not support file locking as it uses a particular data virtualization service (DVS) - see here for documentation of this issue: https://docs.nersc.gov/development/languages/python/parallel-python/#parallel-io-with-h5py. I solved my problem by rolling back to version 22.5.11, right before the file lock was introduced. 

Other filesystems may suffer from similar restrictions. Is it worth adding a setting for cppimport that enables / disables the file lock for multiprocessing build support? For what it's worth, I'm using cppimport in code called from multiple processes, and I protect the build process manually outside the package. I really like the fix introduced by the lock and think it would address the need on most filesystems - but it would be nice to turn it off.

If there is interest in adding the setting, I can write a fix and propose a PR. Feedback solicited!

### PR DESCRIPTION
**Pull request by vbharadwaj-bk**
Fixes #90 

This PR adds the `use_filelock` option to cppimport's global settings (`True` by default). By setting this to `False`, users are no longer protected from multiple processes competing to rebuild an extension. This is necessary on filesystems that don't support file locking, e.g. [here].(https://docs.nersc.gov/performance/io/dvs/#:~:text=DVS%20doesn't%20support%20file,locking%2C%20please%20use%20Perlmutter%20Scratch.)

The README has been updated to describe this behavior. In `test_cppimport.py`, all tests EXCEPT the multiprocessing test have been updated to disable the file lock. The multiprocessing test only runs with the invocation `pytest --multiprocessing`, and it does not disable the file lock to test this feature.

Verified that all tests passing (2 warnings). Feedback welcome!

```
====================== 17 passed, 2 warnings in 42.86s =======================
```

### CODE REVIEW

File: README.md
Comment (tbenthompson) [inline README.md:178]: Mind giving an example or two so people might have a better guess of whether their OS supports locking or not?

**Review by tbenthompson — APPROVED** [review]
This looks wonderful! Thank you so much. 

I left one tiny comment if you don't mind making a small edit to the new readme section.

**Comment by tbenthompson** [comment]
Also looks like the CI flake8 checks are failing on some of the new code. There's a pre-commit config that should get things set up correctly to avoid those issues. but you're also welcome to just flake8 directly.

**Comment by tbenthompson** [comment]
I also invited you as a collaborator. Once the CI is passing, feel free to merge this when you're ready.

**Comment by vbharadwaj-bk** [comment]
No problem, fixed and passed linting. The CI is still failing for Python 3.7; it turns out Mako doesn't build on Python 3.7 since Nov. 9 2023 [see here](https://github.com/sqlalchemy/mako/issues/381) due to a change in jsonscheme (see [here](https://stackoverflow.com/questions/59216175/importerror-cannot-import-name-metadata-from-importlib)).

Should Python 3.7 be removed from the CI?

**Comment by tbenthompson** [comment]
Thanks! Yeah, it seems like we should just remove Python 3.7 from the CI.

**Pull request merged**
